src/Assistant.ts

```diff
@@ -0,0 +1,376 @@
+    this.threadStarted = Array.isArray(threadStarted) ? threadStarted : [threadStarted];
+    this.threadContextChanged = Array.isArray(threadContextChanged) ? threadContextChanged : [threadContextChanged];
+    this.userMessage = Array.isArray(userMessage) ? userMessage : [userMessage];
+  }
+
+  public getMiddleware(): Middleware<AnyMiddlewareArgs> {
+    return async (args): Promise<void> => {
+      if (isAssistantEvent(args) && matchesConstraints(args)) return this.processEvent(args);
+      return args.next();
+    };
+  }
+
+  private async processEvent(args: AllAssistantMiddlewareArgs): Promise<void> {
+    const { payload } = args;
+    const assistantArgs = prepareAssistantArgs(this.threadContextStore, args);
+    const assistantMiddleware = this.getAssistantMiddleware(payload);
+    return processAssistantMiddleware(assistantArgs, assistantMiddleware);
+  }
+
+  /**
+   * `getAssistantMiddleware()` returns the Assistant instance's middleware
+   */
+  private getAssistantMiddleware(payload: AllAssistantMiddlewareArgs['payload']): AssistantMiddleware {
+    switch (payload.type) {
+      case 'assistant_thread_started':
+        return this.threadStarted;
+      case 'assistant_thread_context_changed':
+        return this.threadContextChanged;
+      case 'message':
+        return this.userMessage;
+      default:
+        return [];
+    }
+  }
+}
+
+/**
+ * `prepareAssistantArgs()` takes the event arguments and:
+ *  1. Removes the next() passed in from App-level middleware processing, thus preventing
+ *  events from continuing down the global middleware chain to subsequent listeners
+ *  2. Adds assistant-specific utilities (i.e., helper methods)
+ * */
+export function prepareAssistantArgs(
+  threadContextStore: AssistantThreadContextStore,
+  args: AllAssistantMiddlewareArgs<AssistantMiddlewareArgs>,
+): AllAssistantMiddlewareArgs {
+  const { next: _next, ...assistantArgs } = args;
+  const preparedArgs = { ...assistantArgs as Exclude<AllAssistantMiddlewareArgs<AssistantMiddlewareArgs>, 'next'> };
+
+  // Do not pass preparedArgs (ie, do not add utilities to get/save)
+  preparedArgs.getThreadContext = () => threadContextStore.get(args);
+  preparedArgs.saveThreadContext = () => threadContextStore.save(args);
+
+  preparedArgs.say = createSay(preparedArgs);
+  preparedArgs.setStatus = createSetStatus(preparedArgs);
+  preparedArgs.setSuggestedPrompts = createSetSuggestedPrompts(preparedArgs);
+  preparedArgs.setTitle = createSetTitle(preparedArgs);
+  return preparedArgs;
+}
+
+/**
+ * `isAssistantEvent()` determines if incoming event is a supported
+ * Assistant event type.
+ */
+export function isAssistantEvent(args: AnyMiddlewareArgs): boolean {
+  return ASSISTANT_PAYLOAD_TYPES.has(args.payload.type);
+}
+
+/**
+ * `matchesConstraints()` determines if the incoming event payload
+ * is related to the Assistant.
+ */
+export function matchesConstraints(args: AnyMiddlewareArgs): args is AllAssistantMiddlewareArgs {
```

Comment:
I don't think this type guard is correct 🤔  if the arguments can be any middleware arguments (including non-assistant ones), then the logic below states that any non-message middleware args imply the args are `AllAssistantMiddlewareArgs`. So e.g. if I passed in a `SlackShortcutMiddlewareArgs` into this method, which would have an `args.payload.type` of either `shortcut` or `message_action`, this would return `true`.

Reply:
It's almost certainly not correct and this is just what was required to get by. 🥲 I'll make another attempt to tighten it up.

Reply:
If this method is only used in the context of assistant stuff, then perhaps the type of `args` passed in should not be `AnyMiddlewareArgs` but rather `AnyAssistantMiddlewareArgs`? Perhaps something like this (seems to work locally for me):

```diff
diff --git a/src/Assistant.ts b/src/Assistant.ts
index 13d7b0be..f7f23966 100644
--- a/src/Assistant.ts
+++ b/src/Assistant.ts
@@ -185,7 +185,7 @@ export function prepareAssistantArgs(
  * `isAssistantEvent()` determines if incoming event is a supported
  * Assistant event type.
  */
-export function isAssistantEvent(args: AnyMiddlewareArgs): boolean {
+export function isAssistantEvent(args: AnyMiddlewareArgs): args is AssistantMiddlewareArgs {
   return ASSISTANT_PAYLOAD_TYPES.has(args.payload.type);
 }
 
@@ -193,7 +193,7 @@ export function isAssistantEvent(args: AnyMiddlewareArgs): boolean {
  * `matchesConstraints()` determines if the incoming event payload
  * is related to the Assistant.
  */
-export function matchesConstraints(args: AnyMiddlewareArgs): args is AllAssistantMiddlewareArgs {
+export function matchesConstraints(args: AssistantMiddlewareArgs): args is AllAssistantMiddlewareArgs {
   return args.payload.type === 'message' ? isAssistantMessage(args.payload) : true;
 }
```

Then when these two methods are used in `getMiddleware`, the first use of `isAssistantEvent` narrows the `args` down to `AssistantMiddlewareArgs`, at which point we can pass those into `matchesConstraints`, which only accepts `AssistantMiddlewareArgs`. That makes sense to me in my head but admittedly I have a simplistic/outsider view of this stuff.

Reply:
Good catch. I used your change and made an additional tweak and it appears to work as expected. Thanks for suggesting this!